play-services-core/src/main/java/org/microg/gms/gcm/McsService.java

```diff
@@ -485,6 +507,24 @@ private void handleAppMessage(DataMessageStanza msg) {
             for (ResolveInfo resolveInfo : infos) {
                 logd("Target: " + resolveInfo);
                 Intent targetIntent = new Intent(intent);
+                if(mDeviceIdleController != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+                    try {
+                        java.lang.reflect.Method method = android.os.UserHandle.class.getMethod("getUserId", int.class);
+                        int userId = (int) method.invoke(null, getPackageManager().getApplicationInfo(msg.category, 0).uid);
+                        logd("Adding app " + msg.category + " for userId " + userId + " to the temp whitelist");
+                        mDeviceIdleController.addPowerSaveTempWhitelistApp(resolveInfo.activityInfo.packageName, 10000, userId, "GCM Push");
```

Comment:
As the only reason for importing the hidden AIDL from AOSP, is to call a single method, why not use reflection for the invocation (as you use reflection anyway) and thus get rid of the whole AIDL?

Reply:
This is my fault since this PR is based on my commit and my knowledge of reflection is minimal.

Reply:
Isn't using an AIDL generally preferred over reflection?  The reflection used here is only done because `@hide` was used in the Android API.  Shouldn't we limit our use of reflection to only scenarios that absolutely require it?  I can swap the AIDL for reflection, but I just wanted to make sure this is really what is desired.

(As a side note: I think we might even be able to [avoid using reflection completely](https://stackoverflow.com/questions/7888191/how-do-i-build-the-android-sdk-with-hidden-and-internal-apis-available), but it would require compiling against a modified `android.jar` instead of the standard SDK.  This would make the code a bit cleaner, but the build process a bit more complicated.  So far it doesn't seem like microG needs to access the hidden API very often, but if that changes as more features are implemented, this is probably one of the rare cases that it would be acceptable since we need to mimic Google.)

Reply:
@voidstarstar the problem with importing the AIDL directly, is that it is less likely to survive an android update than reflection. Remember that hidden APIs might be changed without prior notice even in minor updates. Also, once you have an incompatibility between android versions, it's way easier to handle it with reflection than having two AIDL files for the same interface name.

Reply:
If you want to explore the path of using an extended standard library (modified `android.jar`) check out the `unifiednlp-compat` module and its usage in `unifiednlp-base`. This is likely to work here as well, but you have to be careful with cross version compatibility again.